dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp

```diff
@@ -1,43 +1,244 @@
+    String description;
+    WriteBufferFromString buffer(description);
+
+    buffer << "Insertion status:\n";
+
+    auto writeDescription = [&buffer](const std::string & address, size_t shard_id, size_t blocks_wrote)
+    {
+        buffer << "Wrote " << blocks_wrote << " blocks on shard " << shard_id << " replica ";
+        buffer << unescapeForFileName(address) << '\n';
+    };
+
+    size_t job_id = 0;
+    for (size_t shard_id : ext::range(0, shards_info.size()))
+    {
+        const auto & shard_info = shards_info[shard_id];
+        const auto & local_addresses = shard_info.local_addresses;
+
+        for (const auto & address : local_addresses)
+        {
+            writeDescription(address.toStringFull(), shard_id, blocks_inserted + (finished_local_nodes_count ? 1 : 0));
+            if (finished_local_nodes_count)
+                --finished_local_nodes_count;
+        }
+
+        for (const auto & dir_name : shard_info.dir_names)
+        {
+            writeDescription(dir_name, shard_id, blocks_inserted + (done_jobs[job_id] ? 1 : 0));
+            ++job_id;
+        }
+    }
+
+    return description;
+}
+
+void DistributedBlockOutputStream::calculateRemoteJobsCount()
+{
+    remote_jobs_count = 0;
+    const auto & shards_info = cluster->getShardsInfo();
+    for (const auto & shard_info : shards_info)
+        remote_jobs_count += shard_info.dir_names.size();
+}
+
+void DistributedBlockOutputStream::writeSync(const Block & block)
+{
+    if (!pool)
+    {
+        /// Deferred initialization. Only for sync insertion.
+        calculateRemoteJobsCount();
+        pool.emplace(remote_jobs_count);
+    }
+
+    std::vector<bool> done_jobs(remote_jobs_count, false);
+    std::atomic<unsigned> finished_jobs_count(0);
+    std::mutex mutex;
+    std::condition_variable cond_var;
+
+    const Cluster::ShardsInfo & shards_info = cluster->getShardsInfo();
+    Blocks blocks = shards_info.size() > 1 ? splitBlock(block) : Blocks({block});
+
+    size_t job_id = 0;
+    for (size_t shard_id : ext::range(0, blocks.size()))
+    {
+        for (size_t replica_id: ext::range(0, shards_info[shard_id].dir_names.size()))
+        {
+            pool->schedule(createWritingJob(done_jobs, finished_jobs_count, cond_var,
+                                            blocks[shard_id], job_id, shards_info[shard_id], replica_id));
+            ++job_id;
+        }
+    }
+
+    const size_t jobs_count = job_id;
+    size_t finished_local_nodes_count;
+    const auto time_point = deadline;
+    auto timeout = insert_timeout;
+    auto & pool = this->pool;
+    auto wait = [& mutex, & cond_var, time_point, & finished_jobs_count, jobs_count, & pool, timeout]()
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        auto cond = [& finished_jobs_count, jobs_count] { return finished_jobs_count == jobs_count; };
+        if (timeout)
+        {
+            if (!cond_var.wait_until(lock, time_point, cond))
```

Comment:
See "spurious wakeups" in documentation:
http://en.cppreference.com/w/cpp/thread/condition_variable/wait_until